### PR TITLE
core: don't check any parts of a computed set in InstanceDiff.Same

### LINKED ISSUE #2027: aws_security_group - diffs didn't match
Received the following error message:

aws_security_group.nat: diffs didn't match during apply. This is a bug with Terraform and should be reported.

Not sure that this is enough info to reproduce, and clearly there's some issue. 

```
resource "aws_security_group" "nat" {
	vpc_id = "${aws_vpc.main.id}"
	name = "nat"
	description = "For nat servers "

	ingress {
		from_port = 22
		to_port = 22
		protocol = "tcp"
		security_groups = ["${aws_security_group.jump.id}"]
	}

	ingress {
		from_port = 22
		to_port = 22
		protocol = "tcp"
		security_groups = ["${aws_security_group.provision.id}"]
	}

	ingress {
		from_port = 23
		to_port =  65535
		protocol = "tcp"
		security_groups = ["${aws_security_group.ssh_base.id}"]
	}

	ingress {
		from_port = -1
		to_port = -1
		protocol = "icmp"
		security_groups = ["${aws_security_group.ssh_base.id}"]
	}

	ingress {
		from_port = 23
		to_port =  65535
		protocol = "udp"
		security_groups = ["${aws_security_group.ssh_base.id}"]
	}

	egress {
		from_port = 0
		to_port = 0
		protocol = "-1"
		cidr_blocks = ["0.0.0.0/0"]
	}
}
```

### PR DESCRIPTION
**Pull request by jbardin**
When checking for "same" values in a computed hash, not only might some
of the values differ between versions changing the hash, but there may be
fields not included at all in the original map, and different overall
counts.

Instead of trying to match individual set fields with different hashes,
remove any hashed key longer than the computed key with the same base
name.

Fixes #2027 (and hopefully a number of other "diffs didn't match during apply" issues)

### CODE REVIEW

File: terraform/diff.go [hunk 1]
```diff
@@ -443,35 +443,30 @@ func (d *InstanceDiff) Same(d2 *InstanceDiff) (bool, string) {
 			// values. So check if there is an approximate hash in the key
 			// and if so, try to match the key.
 			if strings.Contains(k, "~") {
-				// TODO (SvH): There should be a better way to do this...
 				parts := strings.Split(k, ".")
-				parts2 := strings.Split(k, ".")
+				parts2 := append([]string(nil), parts...)
```
Comment (phinze) [inline terraform/diff.go]: The `append` with `[]string(nil)` base is a new one on me - looks like it makes sure parts2 is a freshly allocated slice?

File: terraform/diff.go [hunk 2]
```diff
@@ -443,35 +443,30 @@ func (d *InstanceDiff) Same(d2 *InstanceDiff) (bool, string) {
 			// values. So check if there is an approximate hash in the key
 			// and if so, try to match the key.
 			if strings.Contains(k, "~") {
-				// TODO (SvH): There should be a better way to do this...
```
Comment (phinze) [inline terraform/diff.go]: haha -> @jbardin responds... "there is"!

File: terraform/diff.go (hunk 1 above)
Comment (jbardin) [inline terraform/diff.go]: yup!  equivalent to  `parts2 := make([]string, len(parts)); copy(parts2, parts)`

**Comment by phinze** [comment]
After chatting on phone about detail - this LGTM!

**Pull request merged**